File: frontend/src/types/portfolio.ts
export type TradingStyle = "short" | "swing" | "long";
export type PortfolioBrokerTemplate =
  | "auto"
  | "generic_cn"
  | "tonghuashun"
  | "guotai_haitong"
  | "eastmoney"
  | "huatai"
  | "pingan";

export interface PortfolioAccountCreate {
  name: string;
  available_funds: number;
  enabled: boolean;
}

export interface PortfolioAccountUpdate {
  name?: string | null;
  available_funds?: number | null;
  enabled?: boolean | null;
}

export interface PortfolioAccountRecord {
  id: string;
  name: string;
  available_funds: number;
  enabled: boolean;
  created_at: string;
  updated_at: string;
}

export interface PortfolioPositionLot {
  acquired_at?: string | null;
  quantity: number;
  cost_price: number;
  note?: string | null;
}

export interface PortfolioPositionCreate {
  account_id: string;
  symbol: string;
  name?: string | null;
  cost_price: number;
  quantity: number;
  available_quantity?: number | null;
  frozen_quantity?: number | null;
  invested_amount?: number | null;
  trading_style: TradingStyle;
  industry?: string | null;
  note?: string | null;
  lots?: PortfolioPositionLot[];
}

export interface PortfolioPositionUpdate {
  account_id?: string | null;
  symbol?: string | null;
  name?: string | null;
  cost_price?: number | null;
  quantity?: number | null;
  available_quantity?: number | null;
  frozen_quantity?: number | null;
  invested_amount?: number | null;
  trading_style?: TradingStyle | null;
  industry?: string | null;
  note?: string | null;
  lots?: PortfolioPositionLot[] | null;
}

export interface PortfolioPositionRecord {
  id: string;
  account_id: string;
  symbol: string;
  name: string;
  cost_price: number;
  quantity: number;
  available_quantity: number;
  frozen_quantity: number;
  invested_amount?: number | null;
  trading_style: TradingStyle;
  industry?: string | null;
  note?: string | null;
  lots: PortfolioPositionLot[];
  created_at: string;
  updated_at: string;
}

export interface PortfolioPositionAdvice {
  headline: string;
  risk: string;
  action: string;
  template: string;
}

export interface PortfolioPositionView extends PortfolioPositionRecord {
  account_name: string;
  latest_price?: number | null;
  change_pct?: number | null;
  market_value?: number | null;
  cost_value: number;
  pnl?: number | null;
  pnl_pct?: number | null;
  daily_pnl?: number | null;
  weight_pct: number;
  available_ratio_pct: number;
  quote_error?: string | null;
  advice?: PortfolioPositionAdvice | null;
}

export interface PortfolioIndustryExposure {
  industry: string;
  market_value: number;
  weight_pct: number;
  position_count: number;
  symbols: string[];
}

export interface PortfolioAccountView extends PortfolioAccountRecord {
  positions: PortfolioPositionView[];
  total_cost: number;
  total_market_value: number;
  total_pnl: number;
  total_pnl_pct: number;
  total_daily_pnl: number;
  total_assets: number;
  position_ratio_pct: number;
  industry_exposures: PortfolioIndustryExposure[];
}

export interface PortfolioMarketSchedule {
  calendar_source: string;
  market_phase: "open" | "closed";
  is_trading_day: boolean;
  next_open_at?: string | null;
  next_refresh_in_ms: number;
}

export interface PortfolioSummary {
  accounts: PortfolioAccountView[];
  available_funds_total: number;
  total_cost: number;
  total_market_value: number;
  total_assets: number;
  total_position_ratio_pct: number;
  total_pnl: number;
  total_pnl_pct: number;
  total_daily_pnl: number;
  quote_error_count: number;
  industry_exposures: PortfolioIndustryExposure[];
  market_schedule: PortfolioMarketSchedule;
}

export interface PortfolioScreenshotImportRequest {
  account_id: string;
  image_data_url?: string | null;
  dry_run?: boolean;
  skip_zero_quantity?: boolean;
  broker_template?: PortfolioBrokerTemplate;
  broker_name?: string | null;
  parsed_rows?: PortfolioScreenshotImportRow[];
}

export interface PortfolioScreenshotImportRow {
  name: string;
  symbol?: string | null;
  quantity: number;
  available_quantity?: number | null;
  cost_price: number;
  latest_price?: number | null;
  market_value?: number | null;
  pnl_amount?: number | null;
  pnl_pct?: number | null;
  action: "created" | "updated" | "skipped" | "preview";
  reason?: string | null;
  position_id?: string | null;
}

export interface PortfolioScreenshotImportResponse {
  account_id: string;
  account_name: string;
  broker_name?: string | null;
  detected_at: string;
  parsed_count: number;
  imported_count: number;
  updated_count: number;
  skipped_count: number;
  rows: PortfolioScreenshotImportRow[];
}

export interface PortfolioCsvImportRequest {
  account_id: string;
  csv_text: string;
  dry_run?: boolean;
  preserve_existing_note?: boolean;
  default_trading_style?: TradingStyle;
}

export interface PortfolioCsvImportRow {
  symbol: string;
  name?: string | null;
  quantity: number;
  available_quantity?: number | null;
  frozen_quantity?: number | null;
  cost_price: number;
  industry?: string | null;
  trading_style: TradingStyle;
  lots: PortfolioPositionLot[];
  action: "created" | "updated" | "skipped" | "preview";
  reason?: string | null;
  position_id?: string | null;
}

export interface PortfolioCsvImportResponse {
  account_id: string;
  account_name: string;
  detected_at: string;
  parsed_count: number;
  imported_count: number;
  updated_count: number;
  skipped_count: number;
  rows: PortfolioCsvImportRow[];
}

File: backend/app/schemas/portfolio.py
from datetime import datetime
from typing import List, Literal, Optional

from pydantic import Field

from .common import ContractModel


TradingStyle = Literal["short", "swing", "long"]
PortfolioBrokerTemplate = Literal[
    "auto",
    "generic_cn",
    "tonghuashun",
    "guotai_haitong",
    "eastmoney",
    "huatai",
    "pingan",
]


class PortfolioAccountCreate(ContractModel):
    name: str = Field(..., min_length=1)
    available_funds: float = 0
    enabled: bool = True


class PortfolioAccountUpdate(ContractModel):
    name: Optional[str] = None
    available_funds: Optional[float] = None
    enabled: Optional[bool] = None


class PortfolioAccountRecord(ContractModel):
    id: str
    name: str
    available_funds: float = 0
    enabled: bool = True
    created_at: datetime
    updated_at: datetime


class PortfolioPositionLot(ContractModel):
    acquired_at: Optional[datetime] = None
    quantity: int = Field(..., gt=0)
    cost_price: float = Field(..., gt=0)
    note: Optional[str] = None


class PortfolioPositionCreate(ContractModel):
    account_id: str
    symbol: str
    name: Optional[str] = None
    cost_price: float = Field(..., gt=0)
    quantity: int = Field(..., gt=0)
    available_quantity: Optional[int] = Field(default=None, ge=0)
    frozen_quantity: Optional[int] = Field(default=None, ge=0)
    invested_amount: Optional[float] = None
    trading_style: TradingStyle = "swing"
    industry: Optional[str] = None
    note: Optional[str] = None
    lots: List[PortfolioPositionLot] = Field(default_factory=list)


class PortfolioPositionUpdate(ContractModel):
    account_id: Optional[str] = None
    symbol: Optional[str] = None
    name: Optional[str] = None
    cost_price: Optional[float] = Field(default=None, gt=0)
    quantity: Optional[int] = Field(default=None, gt=0)
    available_quantity: Optional[int] = Field(default=None, ge=0)
    frozen_quantity: Optional[int] = Field(default=None, ge=0)
    invested_amount: Optional[float] = None
    trading_style: Optional[TradingStyle] = None
    industry: Optional[str] = None
    note: Optional[str] = None
    lots: Optional[List[PortfolioPositionLot]] = None


class PortfolioPositionRecord(ContractModel):
    id: str
    account_id: str
    symbol: str
    name: str
    cost_price: float
    quantity: int
    available_quantity: int = 0
    frozen_quantity: int = 0
    invested_amount: Optional[float] = None
    trading_style: TradingStyle = "swing"
    industry: Optional[str] = None
    note: Optional[str] = None
    lots: List[PortfolioPositionLot] = Field(default_factory=list)
    created_at: datetime
    updated_at: datetime


class PortfolioPositionAdvice(ContractModel):
    headline: str
    risk: str
    action: str
    template: str


class PortfolioPositionView(PortfolioPositionRecord):
    account_name: str
    latest_price: Optional[float] = None
    change_pct: Optional[float] = None
    market_value: Optional[float] = None
    cost_value: float
    pnl: Optional[float] = None
    pnl_pct: Optional[float] = None
    daily_pnl: Optional[float] = None
    weight_pct: float = 0
    available_ratio_pct: float = 0
    quote_error: Optional[str] = None
    advice: Optional[PortfolioPositionAdvice] = None


class PortfolioMarketSchedule(ContractModel):
    calendar_source: str
    market_phase: Literal["open", "closed"] = "closed"
    is_trading_day: bool = False
    next_open_at: Optional[datetime] = None
    next_refresh_in_ms: int = 10000


class PortfolioIndustryExposure(ContractModel):
    industry: str
    market_value: float = 0
    weight_pct: float = 0
    position_count: int = 0
    symbols: List[str] = Field(default_factory=list)


class PortfolioAccountView(PortfolioAccountRecord):
    positions: List[PortfolioPositionView] = Field(default_factory=list)
    total_cost: float = 0
    total_market_value: float = 0
    total_pnl: float = 0
    total_pnl_pct: float = 0
    total_daily_pnl: float = 0
    total_assets: float = 0
    position_ratio_pct: float = 0
    industry_exposures: List[PortfolioIndustryExposure] = Field(default_factory=list)


class PortfolioSummary(ContractModel):
    accounts: List[PortfolioAccountView] = Field(default_factory=list)
    available_funds_total: float = 0
    total_cost: float = 0
    total_market_value: float = 0
    total_assets: float = 0
    total_position_ratio_pct: float = 0
    total_pnl: float = 0
    total_pnl_pct: float = 0
    total_daily_pnl: float = 0
    quote_error_count: int = 0
    industry_exposures: List[PortfolioIndustryExposure] = Field(default_factory=list)
    market_schedule: PortfolioMarketSchedule


class PortfolioScreenshotImportRequest(ContractModel):
    account_id: str
    image_data_url: Optional[str] = None
    dry_run: bool = False
    skip_zero_quantity: bool = True
    broker_template: PortfolioBrokerTemplate = "auto"
    broker_name: Optional[str] = None
    parsed_rows: List["PortfolioScreenshotImportRow"] = Field(default_factory=list)


class PortfolioScreenshotImportRow(ContractModel):
    name: str
    symbol: Optional[str] = None
    quantity: int
    available_quantity: Optional[int] = None
    cost_price: float
    latest_price: Optional[float] = None
    market_value: Optional[float] = None
    pnl_amount: Optional[float] = None
    pnl_pct: Optional[float] = None
    action: Literal["created", "updated", "skipped", "preview"]
    reason: Optional[str] = None
    position_id: Optional[str] = None


class PortfolioScreenshotImportResponse(ContractModel):
    account_id: str
    account_name: str
    broker_name: Optional[str] = None
    detected_at: datetime
    parsed_count: int = 0
    imported_count: int = 0
    updated_count: int = 0
    skipped_count: int = 0
    rows: List[PortfolioScreenshotImportRow] = Field(default_factory=list)


class PortfolioCsvImportRequest(ContractModel):
    account_id: str
    csv_text: str
    dry_run: bool = False
    preserve_existing_note: bool = True
    default_trading_style: TradingStyle = "swing"


class PortfolioCsvImportRow(ContractModel):
    symbol: str
    name: Optional[str] = None
    quantity: int
    available_quantity: Optional[int] = None
    frozen_quantity: Optional[int] = None
    cost_price: float
    industry: Optional[str] = None
    trading_style: TradingStyle = "swing"
    lots: List[PortfolioPositionLot] = Field(default_factory=list)
    action: Literal["created", "updated", "skipped", "preview"]
    reason: Optional[str] = None
    position_id: Optional[str] = None


class PortfolioCsvImportResponse(ContractModel):
    account_id: str
    account_name: str
    detected_at: datetime
    parsed_count: int = 0
    imported_count: int = 0
    updated_count: int = 0
    skipped_count: int = 0
    rows: List[PortfolioCsvImportRow] = Field(default_factory=list)


PortfolioScreenshotImportRequest.model_rebuild()
